File: Number of Paths with Max Score/Number of Paths with Max Score.java
// Solution for Number of Paths with Max Score in JAVA

class Solution {
    int n;
    static final int MOD = 1_000_000_007;
    int[][] scoreMemo, pathMemo;

    private int digit(char ch) {
        return ch == 'S' ? 0 : ch - '0';
    }

    private boolean valid(int i, int j, List<String> board) {
        return i >= 0 && j >= 0 && i < n && j < n && board.get(i).charAt(j) != 'X';
    }

    private int[] solve(int i, int j, List<String> board) {
        char ch = board.get(i).charAt(j);

        if (ch == 'E') return new int[]{0, 1};
        if (ch == 'X') return new int[]{0, 0};

        if (scoreMemo[i][j] != -1)
            return new int[]{scoreMemo[i][j], pathMemo[i][j]};

        int[][] dirs = {{-1, 0}, {0, -1}, {-1, -1}};
        int bestScore = 0, bestPaths = 0;

        for (int[] d : dirs) {
            int ni = i + d[0], nj = j + d[1];
            if (!valid(ni, nj, board)) continue;

            int[] res = solve(ni, nj, board);
            int score = res[1] > 0 ? res[0] + digit(ch) : 0;

            if (score > bestScore) {
                bestScore = score;
                bestPaths = res[1];
            } else if (score == bestScore && res[1] > 0) {
                bestPaths = (bestPaths + res[1]) % MOD;
            }
        }

        if (bestPaths == 0) bestScore = 0;

        scoreMemo[i][j] = bestScore;
        pathMemo[i][j] = bestPaths;

        return new int[]{bestScore, bestPaths};
    }

    public int[] pathsWithMaxScore(List<String> board) {
        n = board.size();
        scoreMemo = new int[n][n];
        pathMemo = new int[n][n];

        for (int[] row : scoreMemo)
            Arrays.fill(row, -1);

        return solve(n - 1, n - 1, board);
    }
}

File: Number of Paths with Max Score/Number of Paths with Max Score.py
# Solution for Number of Paths with Max Score in PY

class Solution:
    MOD = 10**9 + 7

    def pathsWithMaxScore(self, board: List[str]) -> List[int]:
        n = len(board)

        @lru_cache(None)
        def solve(i, j):
            if board[i][j] == 'E':
                return (0, 1)
            if board[i][j] == 'X':
                return (0, 0)

            best_score = best_paths = 0

            for ni, nj in ((i - 1, j), (i, j - 1), (i - 1, j - 1)):
                if 0 <= ni < n and 0 <= nj < n and board[ni][nj] != 'X':
                    score, paths = solve(ni, nj)
                    if paths:
                        score += 0 if board[i][j] == 'S' else int(board[i][j])

                    if score > best_score:
                        best_score, best_paths = score, paths
                    elif score == best_score and paths:
                        best_paths = (best_paths + paths) % self.MOD

            if best_paths == 0:
                best_score = 0

            return (best_score, best_paths)

        return list(solve(n - 1, n - 1))
